src/main/java/org/ensime/maven/plugins/ensime/EnsimeConfigGenerator.java

```diff
@@ -295,20 +312,132 @@ private EnsimeModule ensimeProjectsToModule(final List<EnsimeProject> p) {
+          depMgmtDependencies, allDependencies, ENSIME_SCALA_VERSION, DEFAULT_SCALA_VERSION);
+
+      result._2.ifPresent(logMessage -> log.warn(logMessage));
+
+      return result._1;
+  }
+
+
+  /**
+   * Contains the pure functional logic for determining scala version.
+   *
+   * @return a pair with the first element being the determined
+   *         Scala version, and the second element being an optional
+   *         log message.
+   */
+  public static Pair<String, Optional<String>> getScalaVersion(
+      List<org.apache.maven.model.Dependency> directDependencies,
+      List<org.apache.maven.model.Dependency> depMgmtDependencies,
+      Set<Artifact> allDependencies, String ensimeScalaVersion,
+      String defaultScalaVersion) {
+
+      final String scalaLibraryGroupId = "org.scala-lang";
+      final String scalaLibraryArtifactId = "scala-library";
+
+      // Determine scala version, via a number of different methods:
+
+      // If -Densime.scala.version is set, use that
+      if (ensimeScalaVersion != null && !ensimeScalaVersion.trim().equals(""))
+          return new Pair<String, Optional<String>>(ensimeScalaVersion, Optional.empty());
+
+      // Look for a scala library dependency in <dependencies>
+      Optional<String> directVersion = directDependencies.stream()
+          .filter(d ->
+              d.getGroupId().equals(scalaLibraryGroupId) &&
+              d.getArtifactId().equals(scalaLibraryArtifactId))
+          .findFirst()
+          .map(d -> d.getVersion());
+      if (directVersion.isPresent())
+          return new Pair<String, Optional<String>>(directVersion.get(), Optional.empty());
+
+      // Look in <dependencyManagement>
+      Optional<String> depMgmtVersion = depMgmtDependencies.stream()
+          .filter(d ->
+              d.getGroupId().equals(scalaLibraryGroupId) &&
+              d.getArtifactId().equals(scalaLibraryArtifactId))
+          .findFirst()
+          .map(d -> d.getVersion());
+      if (depMgmtVersion.isPresent())
+          return new Pair<String, Optional<String>>(depMgmtVersion.get(), Optional.empty());
+
+      // Look at all dependencies, including transitive dependencies
+      String[] allVersions = allDependencies.stream()
+          .filter(d ->
+              d.getGroupId().equals(scalaLibraryGroupId) &&
+              d.getArtifactId().equals(scalaLibraryArtifactId))
+          .map(d -> d.getVersion())
+          .collect(toSet())
+          .toArray(new String[0]);
+      // If there is only one distinct version, return that
+      if (allVersions.length == 1)
+          return new Pair<String, Optional<String>>(allVersions[0], Optional.empty());
+      // Group by major version
+      Pattern p = Pattern.compile("^((\\d+)\\.(\\d+))\\.(\\d+)$");
+      Map<String, List<String>> depsByMajorVersion = Arrays.stream(allVersions)
+          .collect(groupingBy(v -> {
+              Matcher m = p.matcher(v);
+              if (m.find()) return m.group(1);
```

Comment:
Technically speaking, this is not major version but major + minor version.

According to the semantic versioning here is the template for versioning:

MAJOR.MINOR.FIX

Reply:
Good point -- I was thinking about Scala binary compatibility and treating, say, 2.12 as the major version, but you're right this isn't technically correct per semantic versioning.  Maybe I should call it something else?  Say "scala binary version" for what is now "major" above and something else for "minor" above?

Reply:
Well, we need to revisit this as soon as Dotty is out, but that won't be anytime soon I think :)